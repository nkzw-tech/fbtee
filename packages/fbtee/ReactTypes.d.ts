import {
  FbtConjunction,
  FbtDelimiter,
  FbtWithoutString,
} from './lib/index.d.ts';

enum IntlVariations {
  GENDER_MALE = 1,
  GENDER_FEMALE = 2,
  GENDER_UNKNOWN = 3,
  NUMBER_ONE = 4,
  NUMBER_TWO = 8,
  NUMBER_MANY = 12,
  NUMBER_ZERO = 16,
  NUMBER_FEW = 20,
  NUMBER_OTHER = 24,
  BITMASK_NUMBER = 28,
}

enum GenderConst {
  NOT_A_PERSON = 0,
  FEMALE_SINGULAR = 1,
  MALE_SINGULAR = 2,
  UNKNOWN_SINGULAR = 7,
  UNKNOWN_PLURAL = 11,
}

export type ParamOptions = {
  /**
   * `IntlVariations.GENDER_*` Pass the gender of the parameter for correctly variated text.
   */
  gender?: IntlVariations;
  /**
   * Passing a value of type `number` into the `number` option uses that value as the input for which we determine the [CLDR plural value](http://cldr.unicode.org/index/cldr-spec/plural-rules).
   * You can pass `true` to simply use the parameter value (the same value that replaces the token).
   */
  number?: number | true;
};

export type PronounOptions = {
  /**
   * Whether to capitalize the pronoun in the source string.
   */
  capitalize?: boolean;
  /**
   * Whether to elide the NOT_A_PERSON option in the text variations generated.
   */
  human?: boolean;
};

export type PluralOptions = {
  /**
   * Represents the "few" form of the string (e.g., 2-4 in Russian). Falls back to `many` if not provided.
   */
  few?: string;
  /**
   * Represents the plural form of the string in English. Default is `{singular} + 's'`
   */
  many?: string;
  /**
   * Name of the token where count shows up. (Default: `"number"`)
   */
  name?: string;
  /**
   * `"yes"|"no"|"ifMany"`: Whether to show the `{number}` in the string.
   *
   * > Note that the singular phrase never has a token, but inlines to `1`.
   * ? This is to account for languages like Hebrew for which showing the actual number isn't appropriate
   *
   *  - "no": (DEFAULT) Don't show the count
   *  - "ifMany": Show the count only in plural case
   *  - "yes": Show the count in all cases
   */
  showCount?: 'yes' | 'no' | 'ifMany';
  /**
   * Represents the "two" form of the string (e.g., exactly 2 in Arabic). Falls back to `many` if not provided.
   */
  two?: string;
  /**
   * For overriding the displayed number
   */
  value?: unknown;
  /**
   * Represents the "zero" form of the string (e.g., 0 in Arabic). Falls back to `many` if not provided.
   */
  zero?: string;
};

export type FbtOptions = {
  /**
   * Text author
   */
  author?: string;
  /**
   *  Use a "common" string repository
   *  To use the strings at runtime, there is the fbt.c(...) function call or the <fbt common=true>...</fbt> JSX API.
   *  NOTE: The transform will throw if it encounters a common string not in the map provided.
   *  See: https://facebook.github.io/fbt/docs/common#runtime-api
   */
  common?: boolean;
  /**
   *  Informs [collection](https://facebook.github.io/fbt/docs/collection/) to skip this string (useful for tests/mocks)
   */
  doNotExtract?: boolean;
  /**
   *  (Default: false) FBT normally consolidates whitespace down to one space (' '). Turn this off by setting this to true
   */
  preserveWhitespace?: boolean;
  /**
   * Project to which the text belongs
   */
  project?: string;
  /**
   * IntlVariations.GENDER_*: Pass an [implicit subject](https://facebook.github.io/fbt/docs/implicit_params/) gender to a partially formed text
   */
  subject?: IntlVariations;
};

type FbtOutput = {
  fbt: string;
  params: string;
};

type FbsOutput = {
  fbt: string;
  params: string;
};

type FbtEnumProps = {
  'enum-range': Array<string> | { [enumKey: string]: string };
  key?: string | null;
  value: string;
};

type FbtParamProps = ParamOptions & {
  key?: string | null;
  name: string;
};

type FbtPluralProps = PluralOptions & {
  count: number;
  key?: string | null;
};

type FbtPronounProps = PronounOptions & {
  gender: GenderConst;
  key?: string | null;
  type: PronounType;
};

type FbtNameProps = {
  gender: IntlVariations;
  key?: string | null;
  name: string;
};

type FbtSameParamProps = {
  key?: string | null;
  name: string;
};

type FbtProps = { key?: string | null } & (
  | (FbtOptions & {
      desc: string;
    })
  | { common: true }
);

type FbtListProps = {
  conjunction?: FbtConjunction;
  delimiter?: FbtDelimiter;
  items: Array<FbtWithoutString | React.ReactElement | null | undefined>;
  key?: string | null;
  name: string;
};

declare module 'react' {
  namespace JSX {
    type PropsWithChildren<P> = P & { children?: React.ReactNode | undefined };
    type PropsWithStringChild<P> = P & {
      children?: string | Array<string> | undefined;
    };

    interface IntrinsicElements {
      fbs: PropsWithChildren<FbtProps>;
      'fbs:enum': FbtEnumProps;
      'fbs:name': PropsWithStringChild<FbtNameProps>;
      'fbs:param': PropsWithStringChild<FbtParamProps>;
      'fbs:plural': PropsWithStringChild<FbtPluralProps>;
      'fbs:pronoun': FbtPronounProps;
      'fbs:same-param': FbtSameParamProps;
      fbt: PropsWithChildren<FbtProps>;
      'fbt:enum': FbtEnumProps;
      'fbt:list': FbtListProps;
      'fbt:name': PropsWithChildren<FbtNameProps>;
      'fbt:param': PropsWithChildren<FbtParamProps>;
      'fbt:plural': PropsWithChildren<FbtPluralProps>;
      'fbt:pronoun': FbtPronounProps;
      'fbt:same-param': FbtSameParamProps;
    }
  }
}
