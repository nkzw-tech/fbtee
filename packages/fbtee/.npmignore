.enum_manifest.json
.src_manifest.json
babel-build.config.js
lib-tmp
scripts
src
