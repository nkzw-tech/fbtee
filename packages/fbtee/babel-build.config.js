import babelSyntaxTypescript from '@babel/plugin-syntax-typescript';
import babelFbteeRuntime from '../babel-plugin-fbtee-runtime/lib/index.js';
import babelFbtee from '../babel-plugin-fbtee/lib/index.js';

export default {
  plugins: [
    babelFbtee,
    babelFbteeRuntime,
    [babelSyntaxTypescript, { isTSX: true }],
  ],
};
