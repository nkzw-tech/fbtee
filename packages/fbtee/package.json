{
  "name": "fbtee",
  "version": "1.7.0",
  "description": "The JavaScript & React Internationalization Framework.",
  "keywords": [
    "fbt",
    "globalization",
    "i18n",
    "internationalization",
    "l10n",
    "localization",
    "react",
    "translation"
  ],
  "repository": {
    "type": "git",
    "url": "git+https://github.com/nkzw-tech/fbtee.git",
    "directory": "packages/fbtee"
  },
  "license": "MIT",
  "type": "module",
  "exports": {
    ".": {
      "types": "./lib/index.d.ts",
      "import": "./lib/index.js",
      "react-server": "./lib/index-server.js",
      "default": "./lib/index.js"
    },
    "./ReactTypes": {
      "types": "./ReactTypes.d.ts"
    },
    "./ReactTypes.d.ts": {
      "types": "./ReactTypes.d.ts"
    },
    "./server": {
      "types": "./lib/index-server.d.ts",
      "import": "./lib/index-server.js",
      "default": "./lib/index-server.js"
    }
  },
  "main": "lib/index.js",
  "module": "lib/index.js",
  "types": "lib/index.d.ts",
  "scripts": {
    "build": "pnpm build:babel && pnpm build:prepend && tsdown lib-tmp/index.tsx -d lib --target=node22 --format=esm --clean --dts && tsdown lib-tmp/index-server.tsx -d lib --target=node22 --format=esm --no-clean --dts",
    "build:babel": "babel --delete-dir-on-start --copy-files --config-file ./babel-build.config.js --out-dir=lib-tmp --extensions=.tsx --keep-file-extension --ignore='src/**/__tests__/*.tsx' src",
    "build:fbtee-strings": "pnpm fbtee collect --include-default-strings=false --src src/*.tsx --out Strings.json && ./scripts/rewrite-filepaths.ts",
    "build:prepend": "node -e \"const file = './lib-tmp/list.tsx'; fs.writeFileSync(file, '// @ts-nocheck\\n' + fs.readFileSync(file, 'utf8'));\""
  },
  "dependencies": {
    "invariant": "^2.2.4"
  },
  "peerDependencies": {
    "@nkzw/babel-plugin-fbtee": "workspace:*",
    "react": "^19.0.0"
  }
}
