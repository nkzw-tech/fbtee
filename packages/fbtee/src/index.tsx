import fbsInternal from './fbs.tsx';
import fbtInternal from './fbt.tsx';
import type { FbsAPI, FbtAPI } from './Types.d.ts';

export { default as IntlVariations } from './IntlVariations.tsx';
export { default as setupFbtee } from './setupFbtee.tsx';
export { default as GenderConst } from './GenderConst.tsx';
export { default as FbtTranslations } from './FbtTranslations.tsx';
export { default as FbtResult } from './FbtResult.tsx';
export { default as list, List } from './list.tsx';

export const fbt = fbtInternal as unknown as FbtAPI;
export const fbs = fbsInternal as unknown as FbsAPI;

export type {
  FbtRuntimeInput,
  FbtRuntimeInput as Translation,
} from './Hooks.tsx';
export type { TranslatedString } from './Types.d.ts';
export type { TranslationDictionary } from './FbtTranslations.tsx';
