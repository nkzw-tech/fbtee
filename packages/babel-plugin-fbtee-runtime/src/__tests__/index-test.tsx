import { transformSync } from '@babel/core';
import presetReact from '@babel/preset-react';
import { describe, it } from '@jest/globals';
import fbtee from '@nkzw/babel-plugin-fbtee';
import fbtAutoImport from '@nkzw/babel-plugin-fbtee-auto-import';
import { withFbtImportStatement } from '@nkzw/babel-plugin-fbtee/src/__tests__/FbtTestUtil.tsx';
import fbteeRuntime from '../index.tsx';

const transform = (source: string) =>
  transformSync(source, {
    ast: false,
    plugins: [fbtAutoImport, fbtee, fbteeRuntime],
    presets: [presetReact],
    sourceType: 'module',
  })?.code || '';

describe('Test hash key generation', () => {
  it('should generate hash key for simply string', () => {
    expect(
      transform(
        withFbtImportStatement(`
        fbt('Foo', 'Bar');
      `),
      ),
    ).toMatchSnapshot();
  });

  it('should generate hash key for nested fbts', () => {
    expect(
      transform(
        withFbtImportStatement(
          `<fbt desc="d">
            <fbt:param
              name="two
  lines">
              <b>
                <fbt desc="test">simple</fbt>
              </b>
            </fbt:param>
            test
          </fbt>;`,
        ),
      ),
    ).toMatchSnapshot();
  });

  it('should auto import for <fbt>', () => {
    expect(
      transform(`
        <fbt desc="d">
          <fbt:param
            name="two
lines">
            <b>
              <fbt desc="test">simple</fbt>
            </b>
          </fbt:param>
          test
        </fbt>;`),
    ).toMatchSnapshot();
  });

  it('should auto import for <fbs>', () => {
    expect(
      transform(`<fbs desc="d">
          test
        </fbs>;`),
    ).toMatchSnapshot();
  });
});

describe('Test enum hash keys generation', () => {
  it('should generate single hash key for fbt with enum under regular mode', () => {
    expect(
      transform(
        withFbtImportStatement(
          `fbt('Foo ' + fbt.enum('a', {a: 'A', b: 'B', c: 'C'}), 'Bar');`,
        ),
      ),
    ).toMatchSnapshot();
  });
});

test('Test replacing clear token names with mangled tokens', () => {
  expect(
    transform(
      withFbtImportStatement(
        `<fbt desc="d">
          <b>Your</b>
          friends
          <b>shared</b>
          <fbt:plural
            many="photos"
            showCount="ifMany"
            count={ex1.count}>
            a photo
          </fbt:plural>
        </fbt>;`,
      ),
    ),
  ).toMatchSnapshot();
});
