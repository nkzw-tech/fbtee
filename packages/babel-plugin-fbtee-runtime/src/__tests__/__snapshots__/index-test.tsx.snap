// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test enum hash keys generation should generate single hash key for fbt with enum under regular mode 1`] = `
"import { fbt } from "fbtee";
fbt._({
  "a": "Foo A",
  "b": "Foo B",
  "c": "Foo C"
}, [fbt._enum('a', {
  "a": "A",
  "b": "B",
  "c": "C"
})], {
  hk: "NT3sR"
});"
`;

exports[`Test hash key generation should auto import for <fbs> 1`] = `
"import { fbs } from "fbtee";
fbs._("test", null, {
  hk: "2zCyVJ"
});"
`;

exports[`Test hash key generation should auto import for <fbt> 1`] = `
"import { fbt } from "fbtee";
fbt._("{two lines} test", [fbt._param("two lines", /*#__PURE__*/React.createElement("b", null, fbt._("simple", null, {
  hk: "2pjKFw"
})))], {
  hk: "2xRGl8"
});"
`;

exports[`Test hash key generation should generate hash key for nested fbts 1`] = `
"import { fbt } from "fbtee";
fbt._("{two lines} test", [fbt._param("two lines", /*#__PURE__*/React.createElement("b", null, fbt._("simple", null, {
  hk: "2pjKFw"
})))], {
  hk: "2xRGl8"
});"
`;

exports[`Test hash key generation should generate hash key for simply string 1`] = `
"import { fbt } from "fbtee";
fbt._("Foo", null, {
  hk: "3ktBJ2"
});"
`;

exports[`Test replacing clear token names with mangled tokens 1`] = `
"var fbt_sv_arg_0;
import { fbt } from "fbtee";
fbt_sv_arg_0 = fbt._plural(ex1.count, "number"), fbt._({
  "*": "{=m0} friends {=m2}{number} photos",
  "_1": "{=m0} friends {=m2} a photo"
}, [fbt_sv_arg_0, fbt._implicitParam("=m0", /*#__PURE__*/React.createElement("b", null, fbt._({
  "*": "Your",
  "_1": "Your"
}, [fbt_sv_arg_0], {
  hk: "3AIVHA"
}))), fbt._implicitParam("=m2", /*#__PURE__*/React.createElement("b", null, fbt._({
  "*": "shared",
  "_1": "shared"
}, [fbt_sv_arg_0], {
  hk: "3CHy8o"
})))], {
  hk: "2mDoBt"
});"
`;
